misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/core/BTestRunner.java

```diff
@@ -108,20 +123,62 @@ private static void executeTestFunctions(ProgramFile[] programFiles) {
         for (TesterinaFunction tFunction : testFunctions) {
             boolean isTestPassed = true;
             String errorMessage = null;
-            try {
-                tFunction.invoke();
-            } catch (BallerinaException e) {
-                isTestPassed = false;
-                errorMessage = e.getMessage();
-                outStream.println("test '" + tFunction.getName() + "' failed: " + errorMessage);
-            } catch (Exception e) {
-                isTestPassed = false;
-                errorMessage = e.getMessage();
-                outStream.println("test '" + tFunction.getName() + "' has an error: " + errorMessage);
+
+            // Check whether the test is disabled.
+            if (tFunction.getTesterinaAnnotation().isDisabled()) {
+                tReport.incrementSkipCount();
+                continue;
+            }
+            // Filtering the test functions by group
+            if (groups != null) {
+                // If provided groups matches and the provided flag is to disable groups the function is skipped
+                if ((isGroupAvailable(groups, tFunction.getTesterinaAnnotation().getGroups()) && disableGroups) || (
```

Comment:
can't we improve this? right now it's three for loops right?

Reply:
This is to identify groups, check the sample https://github.com/yasassri/ballerina/blob/testerina_improvements/misc/testerina/README.md#grouping-tests